vpr/src/pack/prepack.cpp

```diff
@@ -1684,3 +1685,35 @@ static void print_chain_starting_points(t_pack_patterns* chain_pattern) {
 
     VTR_LOG("\n");
 }
+
+void alloc_attraction_groups() {
```

Comment:
put this inside the attraction group class. make a default constructor that sets it to invalid/empty. and then put this code in an initialization routine in the class.

Reply:
if you have all the info you need though, just put this routine into the default constructor.

Reply:
if you can return stuff by reference in these routines, do so